File: src/snippets/Dump.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;

public class Program
{
    public class Meeting
    {
        public Guid Id;
        public Meeting(Guid id) { Id = id; }

        public void Load() { }
        public void Save() { }

        public T ReloadAndRetryOnConflict<T>(Func<T> f) { return f(); }

        public void ReloadAndRetryOnConflict(Action f) { }

        public Participant GetParticipant(Guid userId)
        {
            return null;
        }

        public void SetParticipantUpdated(Participant p) { }

        public List<Participant> GetAllParticipants() { return null; }
    }

    public static void RunInBackground(Action a) { }

    public static Task SendChangedEvent(Meeting m, Guid userId) { return null; }

    public static void Main()
    {
        Console.WriteLine("Hello World");
    }

    public class Participant
    {
        public bool IsHandRaised;
        public bool InLobby;
        public bool IsMuted;
        public Guid Id;

        public void SetInLobby(bool b) { }
        public void SetIsHandRaised(bool b) { }
    }

    public class UserToken
    {
        public string Id;
    }

    public Task<bool> DoWork1() { return null; }

    public Task<bool> DoWork2() { return null; }

    public Task<bool> DoWork() { return null; }

    public static IActionResult Ok() { return null; }
    public static IActionResult Accepted() { return null; }

    [HttpPost("{meetingId}/admitUser/{targetUserId}")]
    public IActionResult AdmitUser(Guid meetingId, Guid targetUserId)
    {
        var meeting = new Meeting(meetingId);
        // The data lives in memory, Load just does a deep copy.
        meeting.Load();

        var updated = meeting.ReloadAndRetryOnConflict(() =>
        {
            var participant = meeting.GetParticipant(targetUserId);
            if (participant?.InLobby != true)
            {
                return false;
            }

            participant.SetInLobby(false);

            // Under the hood, Save will acquire the mutex for this meeting,
            // merge in its dirty properties to the master object, and then
            // call Load before returning.
            // If during the merge it finds the participant was updated via
            // another Save since its last Load, it'll throw a conflict
            // exception instead.
            // ReloadAndRetryOnConflict will then call Load to get clean
            // up-to-date state before rerunning the closure.
            meeting.Save();
            return true;
        });

        if (updated)
        {
            RunInBackground(async () =>
            {
                await SendChangedEvent(meeting, targetUserId);
            });
        }

        return Accepted();
    }

    public void SharedHelper(Meeting meeting, Guid userId)
    { }

    public async Task Process(Meeting meeting)
    {
        RunInBackground(async () =>
        {
            var result = await DoWork2();
            meeting.ReloadAndRetryOnConflict(() =>
            {
                // ... Update property Y based on result and save.
            });
        });

        var result = await DoWork1();
        meeting.ReloadAndRetryOnConflict(() =>
        {
            // ... Update property X based on result and save.
        });
    }

    public async Task Process2(Meeting meeting)
    {
        RunInBackground(async () =>
        {
            // NEW:
            var newMeeting = new Meeting(meeting.Id);
            newMeeting.Load();

            var result = await DoWork2();
            newMeeting.ReloadAndRetryOnConflict(() =>
            {
                // ... Update property Y based on result and save.
            });
        });

        var result = await DoWork1();
        meeting.ReloadAndRetryOnConflict(() =>
        {
            // ... Update property X based on result and save.
        });
    }

    public void LowerHands(Meeting meeting)
    {
        var participants = meeting.GetAllParticipants();
        foreach (var participant in participants)
        {
            participant.SetIsHandRaised(false);

            this.SharedHelper(meeting, participant.Id);
        }

        meeting.Save();
    }
}

File: src/snippets/dump.rs
fn lower_hands(mut meeting: Meeting) {
    let participants = meeting.get_all_participants();
    for participant in participants {
        participant.set_is_hand_raised(false);

        shared_helper(&mut meeting, &participant.id);
    }

    meeting.save();
}

/* fn process(mut meeting: Meeting) {
    std::thread::spawn(move || {
        let result = do_work2();
        meeting.reload_and_retry_on_conflict(|meeting| {
            // ... Update property Y based on result and save.
        });
    });

    let result = do_work1();
    meeting.reload_and_retry_on_conflict(|meeting| {
        // ... Update property X based on result and save.
    });
} */

pub fn admit_user(updated: bool, mut meeting: Meeting, target_user_id: String) {
    // ...
    if updated {
        std::thread::spawn(move || {
            send_changed_event(&mut meeting, &target_user_id);
        });
    }
    accepted()
}

pub struct Meeting {
    pub v: Vec<Participant>,
}

impl Meeting {
    pub fn reload_and_retry_on_conflict<F, R>(&mut self, mut f: F) -> R
        where F: FnMut(&mut Meeting) -> R
    {
        f(self)
    }

    pub fn get_all_participants(&mut self) -> &mut [Participant] {
        &mut self.v
    }

    pub fn save(&mut self) {}
}

pub fn do_work2() -> bool {true}
pub fn do_work1() -> bool {true}
fn accepted() {}
fn send_changed_event(_meeting: &mut Meeting, _s: &str) {}
pub fn shared_helper(_meeting: &mut Meeting, _s: &str) {}
pub struct Participant {
    id: String,
}

impl Participant {
    fn set_is_hand_raised(&mut self, _b: bool) {}
}
